src/ess/bifrost/io/sqw.py

```diff
@@ -0,0 +1,424 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2025 Scipp contributors (https://github.com/scipp)
+
+"""SQW output for BIFROST."""
+
+import dataclasses
+
+import numpy as np
+import numpy.typing as npt
+import scipp as sc
+from scippneutron.io import sqw
+from scippneutron.metadata import ESS_SOURCE
+
+from ess.spectroscopy.types import (
+    Beamline,
+    EnergyBins,
+    GravityVector,
+    IncidentEnergyDetector,
+    Measurement,
+    OutFilename,
+    PulsePeriod,
+    SampleRun,
+    SQWBinSizes,
+)
+
+# Projection axes
+# For now, we always project onto the cartesian axes for simplicity.
+_AXIS_U = sc.vector([1, 0, 0], unit="1/angstrom")
+_AXIS_V = sc.vector([0, 1, 0], unit="1/angstrom")
+_AXIS_W = sc.vector([0, 0, 1], unit="1/angstrom")
+
+
+def save_sqw(
+    path: OutFilename,
+    events: IncidentEnergyDetector[SampleRun],
+    *,
+    bin_sizes: SQWBinSizes,
+    energy_bins: EnergyBins,
+    beamline: Beamline[SampleRun],
+    measurement: Measurement[SampleRun],
+    sample: sqw.SqwIXSample,
+    pulse_period: PulsePeriod,
+    gravity: GravityVector,
+) -> None:
+    """Save events recorded at BIFROST to an SQW file.
+
+    The files written by this function can be read by
+    `Horace <https://pace-neutrons.github.io/Horace>`_ for further processing.
+    For details on the file format, see the documentation for Horace as well as
+    the `SQW <https://scipp.github.io/scippneutron/developer/file-formats/sqw.html>`_
+    developer guide in ScippNeutron.
+
+    The input to this function should not yet contain inelastic coordinates like
+    'energy_transfer' or 'sample_table_momentum_transfer'.
+    Instead, ``save_sqw`` histograms the events according to detector number,
+    experiment setting, and incident energy and computes inelastic
+    coordinates on that histogram.
+    This is done for consistency with Horace.
+
+    Note
+    ----
+    This function requires large amounts of memory to construct intermediate
```

Comment:
Does this mean we should implement some processing-in-chunks mechanism at some point?